File: Tree/Binary_Search_Tree_Iterator.py
# Problem Link:- https://leetcode.com/problems/binary-search-tree-iterator/

"""Implement an iterator over a binary search tree (BST). Your iterator will be 
initialized with the root node of a BST.Calling next() will return the next smallest 
number in the BST."""

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class BSTIterator:

    def __init__(self, root):
        self.ans = []
        self.helper(root)
        
    def helper(self,root):
        while root:
            self.ans.append(root)
            root = root.left
        

    def next(self) -> int:
        node = self.ans.pop()
        if node.right:
            self.helper(node.right)
        return node.val        

    def hasNext(self) -> bool:
        return True if len(self.ans) > 0 else False


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: Tree/BinarySearchTreeIterator.cpp
//Problem Link:- https://leetcode.com/problems/binary-search-tree-iterator/

#include<bits/stdc++.h>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class BSTIterator {
public:
    stack<TreeNode*> st;
    BSTIterator(TreeNode* root) {
        TreeNode* temp = root;
        while (root != NULL){
            st.push(root);
            root = root->left;
        }
    }
    
    int next() {
        int res = st.top()->val;
        TreeNode* temp = st.top();
        st.pop();
        if (temp->right != NULL){
            temp = temp->right;
            while (temp != NULL){
                st.push(temp);
                temp = temp->left;
            }
        }
        return res;
    }
    
    bool hasNext() {
        return (st.empty() ? false : true);
    }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */